packages/uniforms/src/ValidatedForm.tsx

```diff
@@ -132,34 +121,24 @@ export function Validated<Base extends typeof BaseForm>(Base: Base) {
         event.stopPropagation();
       }
 
-      const promise: Promise<void> = new Promise((resolve, reject) => {
-        this.setState(
-          () => ({ submitting: true, validate: true }),
-          () => {
-            this.onValidate().then(() => {
-              super.onSubmit().then(resolve, error => {
-                this.setState({ error });
-                reject(error);
-              });
-            }, reject);
-          },
-        );
-      });
+      this.setState({ validate: true });
+
+      const result = this.onValidate().then(error => {
+        if (error !== null) {
+          return Promise.reject(error);
+        }
 
-      promise
-        // `onSubmit` should never reject, so we ignore this rejection.
-        .catch(noop)
-        .then(() => {
-          // It can be already unmounted.
-          if (this.mounted) {
-            // If validation fails, or `super.onSubmit` doesn't touch `submitting`, we need to reset it.
-            this.setState(state =>
-              state.submitting ? { submitting: false } : null,
-            );
-          }
+        // Validation failed (i.e. returned an error), so no error is present
+        // both in the props nor the state.
+        return super.onSubmit().catch(error => {
+          this.setState({ error });
+          throw error;
```

Comment:
Why is this `throw`ed here if we follow up with `.catch(noop)` later on (line 139)?

Reply:
Think of it as of a _re-rejection_ (as in _re-throwing_ an error). Not doing that would effectively swallow `onSubmit` errors.

The `result.catch(noop);` below is there just to suppress the `Unhandled rejection` errors. Every validation error leads to a rejected promise (that `Promise.reject` in line 128) but it is _technically_ handled (i.e. it's already in the state).